File: templates/harness/template/policies/semgrep/ssrf.py
import httpx
from fastapi import FastAPI
from pydantic import BaseModel

from enricher.fetch import fetch_html
from enricher.safe_url import resolve_public_url

app = FastAPI()


class Req(BaseModel):
    url: str


@app.post("/bad")
async def bad(req: Req) -> dict[str, int]:
    async with httpx.AsyncClient() as client:
        # ruleid: shelf.py-ssrf-unvalidated-request
        res = await client.get(req.url)
    return {"status": res.status_code}


@app.post("/bad2")
def bad2(req: Req) -> dict[str, int]:
    # ruleid: shelf.py-ssrf-unvalidated-request
    return {"status": httpx.get(req.url).status_code}


@app.post("/good")
async def good(req: Req) -> dict[str, str]:
    # ok: shelf.py-ssrf-unvalidated-request
    return {"html": await fetch_html(req.url)}


@app.post("/good2")
def good2(req: Req) -> dict[str, int]:
    target = resolve_public_url(req.url)
    # ok: shelf.py-ssrf-unvalidated-request
    return {"status": httpx.get(target.url).status_code}

File: templates/harness/template/policies/semgrep/ssrf.ts
import { Hono } from "hono";

declare function isPublicHost(h: string): boolean;
declare function assertPublicUrl(u: string): string;
const app = new Hono();

app.post("/bad", async (c) => {
  const body = (await c.req.json()) as { url: string };
  // ruleid: shelf.ts-ssrf-unvalidated-fetch
  const res = await fetch(body.url);
  return c.json(await res.json());
});

app.post("/good", async (c) => {
  const body = (await c.req.json()) as { url: string };
  // sanitizer は「値がそこを通る」必要がある。if で弾くだけでは taint は消えない(Semgrep の制約であり設計上の指針でもある)
  const safe = assertPublicUrl(body.url);
  // ok: shelf.ts-ssrf-unvalidated-fetch
  const res = await fetch(safe);
  return c.json(await res.json());
});
